# Salary survey: Data analysis for salary among managers in the world. 

This will be a project based on the CRISP model, this means here will be developed six steps for the analysis:

1.- Business understanding.

2.- Data understanding.

3.- Preparation.

4.- Modeling.

5.- Evaluation.

6.- Deploy.

Maybe here can be used NLP for open questions. 

Will be developed a story about a fictional organization using this data. 


Data recovered from here: https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html

This data is related with salary in many professional fields around the globe. Here are records from managers, systems engineers, librarians, and so on. The example of the questions is the next:

- How old are you?
- What industry do you work in?
- Job title
- If your job title needs additional context, please clarify here:
- What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)
- How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.
- Please indicate the currency
- If "Other," please indicate the currency here:
- If your income needs additional context, please provide it here:
- What country do you work in?
- If you're in the U.S., what state do you work in?
- What city do you work in?
- How many years of professional work experience do you have overall?
- How many years of professional work experience do you have in your field?
- What is your highest level of education completed?
- What is your gender?
- What is your race? (Choose all that apply.)


with this can be started the search and Exploratory Data Analysis

In [11]:
# Imports cell
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
salary_survey_data:pd.DataFrame = pd.read_csv('./data/Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1 (1).csv', sep=',')

column_names = salary_survey_data.columns


In [4]:
salary_survey_data.head()

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                             55,000                                                                                                                                                                                     
1                                             54,600                                                                                                                                                                                     
2                                             34,000                                                                                                                                                                                     
3                                             62,000                                                                                                                                                                                     
4                                             60,000                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                NaN                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

In [5]:
salary_survey_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27922 entries, 0 to 27921
Data columns (total 18 columns):
 #   Column                                                                                                                                                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                --------------  -----  
 0   Timestamp                                                                                                                                                                                                                             27922 non-null  object 
 1   How old are you?                                                                                             

In [6]:
salary_survey_data.shape

(27922, 18)

This dataset have a shape of 27922 rows and 18 columns. 
Six categorical variables: 
- Industry, 
- US state, 
- Education level, 
- Gender
- Currency  
- Race

Race is treated as string, with values separated by commas, because is a multiple choide question variable. Hence, this last variable requires special treatment, like dummy variables. 

Three rank variables:
- Age
- Total years of work
- Field years of work

Six character variables:
- Job title
- Job context
- Other currency
- Income text
- Country
- City

Two numerical variables
- Annual salary
- Additional compensation


We have one last variable: Timestamp. Date and moment when the observation was recorder.

Numerical variables have a trouble: are actually strings, it's necessary to replace them by numbers based on strings

In [7]:
# This for is used to convert string types to int types in salary column.
for i in range(len(salary_survey_data[column_names[5]])):
  if isinstance(salary_survey_data[column_names[5]][i], str):
    salary_survey_data[column_names[5]][i] = float(salary_survey_data[column_names[5]][i].replace(",", ""))

for i in range(len(salary_survey_data[column_names[6]])):
  if isinstance(salary_survey_data[column_names[6]][i], str):
    salary_survey_data[column_names[6]][i] = float(salary_survey_data[column_names[5]][i].replace(",", ""))

# This is used to replace NaN for 0's

salary_survey_data[column_names[5]] = salary_survey_data[column_names[5]].replace(np.nan, 0)
salary_survey_data[column_names[6]] = salary_survey_data[column_names[6]].replace(np.nan, 0)


/tmp/ipykernel_14262/2446780987.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  salary_survey_data[column_names[5]][i] = float(salary_survey_data[column_names[5]][i].replace(",", ""))


In [14]:
for i in range(len(column_names)):
  print(f'column number {i} is {column_names[i]}')

column number 0 is Timestamp
column number 1 is How old are you?
column number 2 is What industry do you work in?
column number 3 is Job title
column number 4 is If your job title needs additional context, please clarify here:
column number 5 is What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)
column number 6 is How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.
column number 7 is Please indicate the currency
column number 8 is If "Other," please indicate the currency here: 
column number 9 is If your income needs additional context, please provide it here:
column number 10 is What country do you work in?
column number 11 is If you're in the U.S., what state do you work i

In [8]:
salary_survey_data.head()

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                            55000.0                                                                                                                                                                                     
1                                            54600.0                                                                                                                                                                                     
2                                            34000.0                                                                                                                                                                                     
3                                            62000.0                                                                                                                                                                                     
4                                            60000.0                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                0.0                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

Now, there are no NaN values. They where replaced by float numbers. Now is necessary to create a correlation matrix. Will be created a new DataFrame with categorical variables and numerical variables, deleting the free answer variables, and categorical variables will have numbers instead categories, just temporary, to create the matrix.

In [ ]:
categorical_variables = [
  column_names[1],
  column_names[2],
  column_names[3],
  column_names[7],
  column_names[10],
  column_names[11],
  column_names[12],
  column_names[13],
  column_names[14],
  column_names[15],
  column_names[16],
  # column_names[17], This have to be transformed again
  ]

continuous_variables = [
  column_names[5],
  column_names[6],
]

# This function is used to replace missing values in cat variables with a -99 
# To indentify null values. 
for cat_var in categorical_variables:
  salary_survey_data[cat_var] = salary_survey_data[cat_var].replace(np.nan, '-99')
